### DISCUSSION TITLE
specific Layout for when screensaver is active

### DISCUSSION THREAD

**Discussion by nico1080** [mattblovell/kodi_panel#44] [Ideas]
It would be nice to have a specific layout for screensaver.

After some digging I found this post on kodi forum
https://forum.kodi.tv/showthread.php?tid=141078

Even if it is old, I confirm that on kodi 18.9 it work
a Json request `{"jsonrpc": "2.0", "method": "XBMC.GetInfoBooleans", "params": { "booleans": ["System.ScreenSaverActive "] }, "id": 1}` will return true or false based on screensaver status

An other solution that use info label instead of booleans is `Window.Property(xmlfile)` that will return this on my libreelec system `/storage/.kodi/addons/screensaver.kaster/resources/skins/default/1080i/screensaver-kaster.xml`. But it will system specific so I believe the boolean method is better.

Additionaly one could also imagine other layouts for when video/music is paused and screen saver become active:

- Idle screensaver layout when screensaver is active i and no media is paused (player active)
- music screensaver layout when screensaver is active and music is paused
- video screensaver layout when screensaver is active and video is paused

More generally a list of all available boolean in kodi:
https://kodi.wiki/view/List_of_boolean_conditions

It could be nice to retrieve it the same way as infolabel and use them is `display_if` or `display_ifnot`
I do not see concrete example yet, but it will come.

**Comment by mattblovell** [comment]
Opened enhancement Issue #45.  Note that a second network call will be necessary (InfoBooleans have to be retrieved separately from InfoLabels), slowing things down somewhat.

**Comment by mattblovell** [comment]
Hmm.... maybe multiple RPC methods can be combined in a single JSON-RPC request. I'm reading about "batch" requests.

I'll have to experiment.

**Comment by mattblovell** [comment]
Should be possible with kodi_panel v1.44

**Comment by nico1080** [comment]
I confirm that booleans retrieve work, and I used for time shift detection: de6eb80fca33938a8b02cc7453a3cd9122f0ce6f

Regarding screensaver detection I  was not able to make it work, I was not able to understand how autoselect work for status screen.

**Comment by mattblovell** [comment]
> I was not able to understand how autoselect work for status screen.

See the text below, which I've added to the example setup TOML files.

The structure is intended to be similar to that used for auto-selection of video screen layouts, audio screen layouts, etc.  The main difference is that the default behavior falls back to the historical one of just using a single-level layout.

The auto-selection function described should be able to make use of the ``System.ScreenSaverActive`` boolean to make a different selection from ``kodi_panel_display.IDisplay``.


Example function
-------------------

In startup script:
```
def my_status_select(info):
    if info['System.ScreenSaverActive']:
        return kodi_panel_display.IDisplay["screensaver"]
    else:
        return kodi_panel_display.IDisplay[config.settings["STATUS_INITIAL"]]

kodi_panel_display.STATUS_SELECT_FUNC = my_status_select
```

where setup.toml is also modified to specify

```
STATUS_NAMES = [ "default", "screensaver" ]
STATUS_INITIAL = "default"
```

and the (now multi-level) ``STATUS_LAYOUT`` table provide elements for both "default" and "screensaver", e.g. defining both

``[STATUS_LAYOUT.default.thumb]``   and
``[STATUS_LAYOUT.screensaver.thumb]``.



New text in example setups
-------------------------------

```
# Status/Info Layout Names
# ------------------------
#
#   This section is entirely optional and thus can be left commented
#   out unless multiple status/info screens are desired.  If one is
#   content with just a single layout (shown either upon a touch
#   interrupt or enabled whenever Kodi is idle), then a single-level
#   layout table further below suffices.  (This becomes a single-level
#   dictionary in Python.)
#
#   An example single-level status layout consists of entries such as
#
#    [STATUS_LAYOUT.thumb]
#      posx = 5
#      posy = 5
#      size = 128
#
#    [[STATUS_LAYOUT.fields]]
#      name = "version"        
#      posx = 145
#      posy = 8
#      font = "font_main"
#      fill = "color_artist"
#
#
#   If multiple status/info layouts are desired, then one MUST
#   define both of the following variables:
#
#      STATUS_NAMES     array of strings specifying the name of
#                       each status/info screen layout
#
#      STATUS_INITIAL   string specifying default layout
#
#   Additionally, one's startup script MUST then define an
#   auto-selection function that decides which layout to use, given
#   the results of retrieving STATUS_LABELS and STATUS_BOOLEANS from
#   Kodi.
#
#   The auto-selection function would be similar in structure to
#   video_select_default() in kodi_panel_display.py.  (One must make
#   use of the kodi_panel_display namespace when making type or
#   variable references, but the structure of the code itself is
#   similar to that example funciton.)
#
#   The user would define the appropriate functionality in whatever
#   startup script is used (e.g., kodi_panel_ili9341 or kodi_panel_fb)
#   and point to it by overriding
#
#     kodi_panel_display.STATUS_SELECT_FUNC
#
#   Finally, the layout tables defined below for STATUS_LAYOUT must
#   define the elements desired for each of the names given in the
#   STATUS_NAMES array.  So, if that array of names is specified
#   as "my_layoutA" and "my_layoutB", then the tables below will
#   end up elements such as
#
#     [STATUS_LAYOUT.my_layoutA.thumb]  and
#     [[STATUS_LAYOUT.my_layoutA.fields]]
#
#   and
#
#     [STATUS_LAYOUT.my_layoutB.thumb]  and
#     [[STATUS_LAYOUT.my_layoutB.fields]]
#
# --------------------------------------------------------------------
# STATUS_NAMES = []
# STATUS_INITIAL = ""
```

**Comment by nico1080** [comment]
It work, I did it in my fork ba5ad2931c2127f4b1036d2d879fc35ab4bbf51a

It also needed `STATUS_AUTOSELECT = true` line in `setup.toml`
